doc/SAI-Proposal-ICMP-ECHO_Extensions.md

```diff
@@ -0,0 +1,415 @@
+#  ICMP ECHO offload
+-------------------------------------------------------------------------------
+ Title       | ICMP ECHO offload
+-------------|-----------------------------------------------------------------
+ Authors     | Chikkegowda Chikkaiah, Cisco
+ Status      | In review
+ Type        | Standards track
+ Created     | 2024-01-1 - Initial Draft
+ SAI-Version | 1.12
+-------------------------------------------------------------------------------
+
+
+## 1.0  Introduction
+
+The traditional use of ICMP-based link prober is to identify link failures. This prober sends ICMP echo packets, receives any ICMP echo replies, and reports the detection results.
+
+When control plane application carries out the link probe (software link failure detection), the system operates at a slow, periodic exchange of probe requests and replies since it uses the kernel network stack to send and receive control packets. This may prolong link failure detection. To decrease this detection interval, the application could utilize the HW/ASIC to send and receive link probe packets at a quicker rate.
+
+The objective of this document is to propose a SAI specification to simplify the ICMP-based probe as a basic service. The service primitives offered by ICMP echo include creating, destroying, and modifying a session, considering the destination address and other parameters. In response, ICMP Echo sends a signal to its clients when the ICMP Echo session starts or stops.
+
```

Comment:
Sure, I can add some details about ICMP reachability consideration over multihop networks.

Regarding differentiating drop and actual link down. The proposal here is same as BFD.  
BFD is designed to be robust against isolated packet drops. Since it sends control packets at regular, short intervals, the loss of a single packet doesn't immediately trigger a link down event. Instead, the link is declared down only if a predetermined number of consecutive packets are missed. This threshold is configurable and can be set based on the expected level of packet loss in the network.

Similar to BFD, ICMP echo also has configurable RX & TX intervals. RX interval (threshold) can be set based on the expected level of packet loss in the network. 

However, it's important to note that BFD itself doesn't explicitly tell you whether a link is down due to a physical failure or due to packet drops; it only indicates that the expected BFD packets have not been received. 
So, in both cases, additional diagnostic steps or strategies may be required to determine the exact cause of the failure.

In summary, HW offloaded session (ASIC level) will just transmit packets periodically and report state of reception of icmp echo packets, as per the configured TX/RX intervals, to the application or client. Beyond this, application can implement, if required, other strategies to make sure its not false link failure.  

Please correct me, if I'm wrong or missed something.